Relay health check v2

dmanjunath opened the pull request.

vicky-g commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -10,35 +10,51 @@ const axios = require('axios')
 
 // Defaults used in relay health check endpoint
-const RELAY_HEALTH_ONE_HOUR_AGO_BLOCKS = 120 // 1 block/5sec = 120 blocks/hr
+const RELAY_HEALTH_TEN_MINS_AGO_BLOCKS = 120 // 1 block/5sec = 120 blocks/10 minutes
```
👍

vicky-g commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -10,35 +10,51 @@ const axios = require('axios')
+  /*
+  There are a few scenarios where a health check should return unhealthy
+  1. Some number of relays are failing for some number of users
+     To solve this, have an object in redis which 
```
is there meant to be more to this comment?

dmanjunath replied:
good catch, fixed

vicky-g commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -76,29 +93,71 @@ module.exports = function (app) {
-            if (!resp.status) failureTxHashes.push(txHash)
+            
+            // tx failed
+            if (!resp.status){
```
nit space after )

dmanjunath replied:
fixed!

vicky-g commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -10,35 +10,51 @@ const axios = require('axios')
 const RELAY_HEALTH_MAX_ERRORS = 5 // max acceptable errors for a 200 response
+const RELAY_HEALTH_MIN_NUM_USERS = 3 //min number of users affected to qualify for error
 const RELAY_HEALTH_MAX_BLOCK_RANGE = 200 // max block range allowed from query params
+const RELAY_HEALTH_SENT_VS_ATTEMPTED_THRESHOLD = 0.2
```
could we have a comment about what this represents? i can't really tell from var name

dmanjunath replied:
yep! added some comments

dmanjunath commented:
@vicky-g comments addressed

vicky-g commented on identity-service/src/txRelay.js:
```diff
@@ -68,6 +68,7 @@ const sendTransactionInternal = async (req, web3, resetNonce = false, txProps, r
     senderAddress,
     gasLimit
   } = txProps
+  const redis = req.app.get('redis')
```
could we add a comment here about the decision to use redis here? :)

dmanjunath replied:
@vicky-g I can if you think it’ll help but this is a pretty common pattern in our code are to use Redis to store metrics/debug info. What would you like the comment to say?

vicky-g replied:
hmmm looking more at the redis functionality, i think its clear enough about what the intention here is to not include a comment explanation. i thought itd be a good idea to mention the reason we're using redis here vs prior to the pr where we weren't using it

SidSethi commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -76,29 +95,70 @@ module.exports = function (app) {
+      flatten(Object.values(failureTxs)).length > maxErrors) isError = true
+
+    // delete old entries from set in redis
+    const epochOneHourAgo = Math.floor(Date.now() / 1000) - 3600
```
looks like the MAX_BLOCK_RANGE = 200, which is ~17min, so is this cache eviction just set to be some amount larger than that? but also i think you have set a ceiling on how old the startBlock can be, so this just needs to be earlier than the earliest possible startBlock?

dmanjunath replied:
so cache eviction is set to all items before 1 hour ago. it is technically possible to set a start and end block override greater than an hour ago. so maybe what we do here is force endBlock to always be the latest block. Max time you can look back is 17 minutes, but it caches for 1 hour

SidSethi replied:
endBlock = latest block fixes a ton of this complexity, sounds good

SidSethi commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -76,29 +95,70 @@ module.exports = function (app) {
+    await redis.zremrangebyscore('relayTxSuccesses', '-inf', epochOneHourAgo)
+
+    // check if there have been any attempts in the time window that we processed the block health check
+    const attemptedTxsInRedis = await redis.zrangebyscore('relayTxAttempts', minBlockTime, '+inf')
```
given that you iterate through range of blocks from startBlock to endBlock, and minBlockTime is the timestamp of the oldest block before startBlock or txCounter is hit, shouldn't this zrangebyscore be from minBlockTime to endBlock.timestamp?

dmanjunath replied:
good point. i'm going to default end block to current block to avoid this issue. also removes cache availability problem for blocks older than 1 hour, since those would have been expired

SidSethi commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -76,29 +95,70 @@ module.exports = function (app) {
+        successfulTxsInRedis,
+        failureTxsInRedis
+      }
+    }
```
nit but maybe blockchain.failedTransactionHashes should also be verbose conditional?

dmanjunath replied:
could you elaborate on this?

SidSethi replied:
discussed offline

SidSethi commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -76,29 +95,70 @@ module.exports = function (app) {
+        successfulTxsCount: successfulTxsInRedis.length,
+        failureTxsCount: failureTxsInRedis.length
+      },
       timeElapsed: Date.now() - start
```
nit - this is slightly confusing, since unclear if timeElapsed during startBlock to endBlock range. maybe healthCheckComputeTime or smth similar?

dmanjunath replied:
done

dmanjunath commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -10,35 +10,53 @@ const axios = require('axios')
 const RELAY_HEALTH_MAX_BLOCK_RANGE = 200 // max block range allowed from query params
+// if (txs on blockchain / attempted) is less than this percent, health check will error
+// eg. 10 transactions on chain but 120 attempts should error
+const RELAY_HEALTH_SENT_VS_ATTEMPTED_THRESHOLD = 0.2
```
@SidSethi we definitely could output this. the reason i picked this number as default (also can be overridden via env var) is because libs retries relays up to 5 times. basically the weight here. is determined by number of attempts taking into account retries and the number of successes. i set it lower rather than higher because in our cases the success txs count is 0, so this check will trigger. and i didn't want it to trigger on false positives if one person is retrying a failing tx over and over

dmanjunath commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -76,29 +95,70 @@ module.exports = function (app) {
+
+            // tx failed
+            if (!resp.status) {
+              const senderAddress = await redis.hget('txHasToSenderAddress', txHash)
```
fixed

dmanjunath commented on identity-service/src/txRelay.js:
```diff
@@ -126,20 +127,33 @@ const sendTransactionInternal = async (req, web3, resetNonce = false, txProps, r
     })
+
     await prom // resolves when a hash exists for a transaction, proving that it has not errored
+    await redis.zadd('relayTxSuccesses', Math.floor(Date.now() / 1000), JSON.stringify(redisLogParams))
```
yeah it's kind of confusing. i added a few comments about how the promise execution is handled

dmanjunath commented on identity-service/src/txRelay.js:
```diff
@@ -126,20 +127,33 @@ const sendTransactionInternal = async (req, web3, resetNonce = false, txProps, r
     receiptPromise = web3.eth.sendSignedTransaction(signedTx)
     const prom = new Promise(function (resolve, reject) {
       let resolved = false
       receiptPromise.once('transactionHash', function (hash) {
```
this is according to the docs. i put a link in the code now

dmanjunath commented on identity-service/src/routes/healthCheck.js:
```diff
@@ -10,35 +10,53 @@ const axios = require('axios')
     // In the case that no query params are defined and the endBlockNumber is less than the max block range, default startBlockNumber to 0
-    const defaultStartBlockNumber = endBlockNumber - RELAY_HEALTH_ONE_HOUR_AGO_BLOCKS >= 0
-      ? endBlockNumber - RELAY_HEALTH_ONE_HOUR_AGO_BLOCKS : 0
+    const defaultStartBlockNumber = endBlockNumber - RELAY_HEALTH_TEN_MINS_AGO_BLOCKS >= 0
```
if i understand you correctly, it's not exactly what you described. endblock is the latest block (unless overridden). startblock is endblock - RELAY_HEALTH_TEN_MINS_AGO_BLOCKS(unless overridden, in which case it needs to be within 200 blocks of end block). the special case here is if you're doing this locally, you can't subtract 120 blocks and go negative. so the check exists to default to 0. not sure where the 200 blocks.

SidSethi replied:
makes sense lol

dmanjunath commented on identity-service/src/txRelay.js:
```diff
@@ -126,20 +127,33 @@ const sendTransactionInternal = async (req, web3, resetNonce = false, txProps, r
     const prom = new Promise(function (resolve, reject) {
       let resolved = false
       receiptPromise.once('transactionHash', function (hash) {
+        redis.hset('txHasToSenderAddress', hash, senderAddress)
```
yep. done

dmanjunath commented:
@SidSethi great review thanks for that. takeaways are 
1. i removed the ability for end block to be overridden. now it's always latest block which removes a lot of complexity around start/end times and also if the cache was pruned the values wouldn't return 0 so you have a div by 0 problem
2. the threshold for sent vs attempted is very much malleable. it's also able to be set via query param so we can tweak it on the fly
3. i added some comments to explain the flow, particularly around the promise rejecting and the promi-event
